include/deal.II/lac/la_parallel_vector.h

```diff
@@ -0,0 +1,1434 @@
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__la_parallel_vector_h
+#define dealii__la_parallel_vector_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/mpi.h>
+#include <deal.II/base/numbers.h>
+#include <deal.II/base/partitioner.h>
+#include <deal.II/base/std_cxx11/shared_ptr.h>
+#include <deal.II/base/thread_management.h>
+#include <deal.II/lac/vector_space_vector.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace parallel
+{
+  namespace distributed
+  {
+    template <typename> class BlockVector;
+  }
+}
+
+namespace LinearAlgebra
+{
+  template <typename> class ReadWriteVector;
+}
+
+#ifdef DEAL_II_WITH_PETSC
+namespace PETScWrappers
+{
+  namespace MPI
+  {
+    class Vector;
+  }
+}
+#endif
+
+#ifdef DEAL_II_WITH_TRILINOS
+namespace TrilinosWrappers
+{
+  namespace MPI
+  {
+    class Vector;
+  }
+}
+#endif
+
+namespace LinearAlgebra
+{
+  namespace distributed
+  {
+    /*! @addtogroup Vectors
+     *@{
+     */
+
+    /**
+     * Implementation of a parallel vector class. The design of this class is
+     * similar to the standard ::dealii::Vector class in deal.II, with the
+     * exception that storage is distributed with MPI.
+     *
+     * The vector is designed for the following scheme of parallel
+     * partitioning:
+     * <ul>
+     * <li> The indices held by individual processes (locally owned part) in
+     * the MPI parallelization form a contiguous range
+     * <code>[my_first_index,my_last_index)</code>.
+     * <li> Ghost indices residing on arbitrary positions of other processors
+     * are allowed. It is in general more efficient if ghost indices are
+     * clustered, since they are stored as a set of intervals. The
+     * communication pattern of the ghost indices is determined when calling
+     * the function <code>reinit (locally_owned, ghost_indices,
+     * communicator)</code>, and retained until the partitioning is changed.
+     * This allows for efficient parallel communication of indices. In
+     * particular, it stores the communication pattern, rather than having to
+     * compute it again for every communication. For more information on ghost
+     * vectors, see also the
+     * @ref GlossGhostedVector "glossary entry on vectors with ghost elements".
+     * <li> Besides the usual global access operator() it is also possible to
+     * access vector entries in the local index space with the function @p
+     * local_element(). Locally owned indices are placed first, [0,
+     * local_size()), and then all ghost indices follow after them
+     * contiguously, [local_size(), local_size()+n_ghost_entries()).
+     * </ul>
+     *
+     * Functions related to parallel functionality:
+     * <ul>
+     * <li> The function <code>compress()</code> goes through the data
```

Comment:
Right now the `compress()` and `update_ghost_values()` methods are still needed. On the one hand because they are as of now still doing the actual data exchange in the two import() methods of `ReadWriteVector` and the current vector.  I plan to address that in a later commit when I switch to lightweight conversion without copy of the locally owned range as discussed in #2535 (this includes also `operator()`, `local_element()`, and `begin()/end()`).